Update mapping logic in `relabel_nodes`

rossbar opened the pull request:
Related to #5903 and #5896.

#5903 was a very nice solution to the problem identified in #5896 - i.e. the input-munging logic for the `mapping` parameter used a `hasattr("__getitem__")` check which was fragile in the case of the built-in `str`. The trick here was that `str` is a Python *type*, which both has a `__getitem__` attr, *and* is a Callable.

The logic has been updated to fix this specific case in #5903, but I'd like to propose using `isinstance` checks against ABC's if for no other reason that I personally find them slightly easier to reason about.

One major difference with this change is how the code fails. Note that previously if a user passed in an object that was neither a `Mapping` nor `Callable`, then they would eventually get an `AttributeError` when the `.get` method was called. For example

#### On main (after #5903)

```python
>>> G = nx.path_graph(4)
>>> s = "0123"
>>> l = ["0", "1", "2", "3"]
>>> nx.relabel_nodes(G, s)
Traceback (most recent call last)
   ...
AttributeError: 'str' object has no attribute 'get'
>>> nx.relabel_nodes(G, l)
Traceback (most recent call last)
   ...
AttributeError: 'list' object has no attribute 'get'
```

#### This PR

```python
>>> G = nx.path_graph(4)
>>> s = "0123"
>>> l = ["0", "1", "2", "3"]
>>> nx.relabel_nodes(G, s)
Traceback (most recent call last)
   ...
TypeError: 'str' object is not callable
>>> nx.relabel_nodes(G, l)
Traceback (most recent call last)
   ...
TypeError: 'list' object is not callable
```

Note the exception changes from an `AttributeError` to a `TypeError` with this proposed change. In principle, the check could be made more specific to improve the exception message, e.g. something like:

```python
if not isinstance(mapping, Mapping):
    if not isinstance(mapping, Callable):
        raise TypeError("mapping must either be a mapping or callable")  # or NetworkXError, if preferred
    m = {n: mapping(n) for n in G}
```

Of course, the change in exception type is a breaking change, so it could well be that this churn isn't worth it. Nevertheless I wanted to submit my two cents on the topic (since I missed the review in #5903!)

dschult commented:
I went through this code and it looks like we expect `mapping` to provide:
`__getitem__, keys, items, values, get, __contains__`.  So the only methods from `abc.Mapping` we are not using are `__iter__, __eq__, __ne__`.  

I have refrained from using the abc classes when we don't use most of the methods required. This case seems like a pretty good match to `Mapping`. 

I worry a little for people who have homemade mappings but have not connected them to the abc.Mapping class. That's a potential backward incompatibility issue.  

Overall this looks good to me! :}

dschult commented on networkx/tests/test_relabel.py:
```diff
@@ -106,12 +106,20 @@ def mapping(n):
         assert nodes_equal(H.nodes(), [65, 66, 67, 68])
 
-    def test_relabel_nodes_classes(self):
+    def test_relabel_nodes_callable_type(self):
         G = nx.empty_graph()
         G.add_edges_from([(0, 1), (0, 2), (1, 2), (2, 3)])
```
Can you take this chance to make this test start with:
`G = nx.path_graph(4)`
Not related to this PR...

rossbar commented:
Thanks for taking a look @dschult - your comments led me down a rabbit hole and, good news!, I learned something new.

I had been under the impression that the *existence of methods* was *always* sufficient to pass an isinstance check, and that explicit inheritance from e.g. an ABC (or a minimal implementation like `UserDict` in the case of a user-defined mapping) wasn't necessarily required. It turns out this is incorrect - the collections.abc docs are very clear on this. In fact, they use `Mapping` as an example of why this isn't the case (see [bullet point number 3 from the docs](https://docs.python.org/3/library/collections.abc.html))!

Fortunately, it turns out that a different approach with the logic that makes the proposed change even simpler. What we really want to test is whether the `mapping` argument is *callable* - if so, apply it to all the nodes to create the mapping. This has the added advantage that it doesn't change the type of the exception raised when a user passes in something that is neither a mapping nor callable.

dschult commented:
That seems good, too.  Technically it does change the current logic if a previous user inputs a class that is both a Mapping and a Callable.  They get the callable behavior instead of the mapping behavior.  Which do we want to "win" when an object can do both?

More rabbit-hole digging: A homemade callable object can register with `Callable` to appear as a subclass of Callable using `Callable.register(obj)`.  I guess we could add to the doc_string that a user can register an object as Callable that way.  But as we convert the `hasattr` logic to `isinstance` throughout the library we will likely need to add such language in many places.

I'm fine with this change. The change in behavior for objects that provide both calls and mappings should probably be noted in the release notes.

dschult commented on networkx/relabel.py:
```diff
@@ -114,13 +116,9 @@ def relabel_nodes(G, mapping, copy=True):
-    ):
+    # you can pass any callable e.g. f(old_label) -> new_label or
+    # e.g. str(old_label) -> new_label, but we'll just make a dictionary here regardless
+    if isinstance(mapping, Callable):
```
More rabbit hole digging.  There is a Python builtin function named `callable` which avoids (or maybe implements) the whole "how do we know if this thing is a class or a function or a method or what?", and it means that people don't have to worry about registering their object as a subclass of `abc.Callable`.
We would still need to note the change in the API due to treatment of objects that are both callable and dict-like.
```python
if callable(mapping): 
```

rossbar replied:
Good catch - that's even better as it removes the import from collections.abc.

rossbar commented:
Okay, I've switched to `callable` and added a release note.

Just to summarize: the main upside here is mostly cosmetic i.e. improving code readability, at the expense subtle changes to the input munging logic. Whether or not the tradeoff is worth it is in the eye of the beholder --- closing this PR and sticking with the status quo would be totally reasonable!

dschult reviewed (approved):
This looks good to me!
